retrieval/scrape.go

```diff
@@ -0,0 +1,243 @@
+package retrieval
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/log"
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/storage"
+)
+
+// Scraper retrieves samples.
+type Scraper interface {
+	// Sends samples to ch until an error occurrs, the context is canceled
+	// or no further samples can be retrieved. The channel is closed before returning.
+	Scrape(ctx context.Context, ch chan<- *model.Sample) error
+}
+
+// Scrape pools runs scrapers for a set of targets where a target, identified by
+// a label set is scraped exactly once.
+type ScrapePool struct {
+	appender storage.SampleAppender
+
+	ctx    context.Context
+	cancel func()
+
+	mtx    sync.RWMutex
+	loops  map[model.Fingerprint]*scrapeLoop
+	states map[model.Fingerprint]*TargetStatus
+}
+
+func NewScrapePool(app storage.SampleAppender) *ScrapePool {
+	sp := &ScrapePool{
+		appender: app,
+		states:   map[model.Fingerprint]*TargetStatus{},
+	}
+	sp.ctx, sp.cancel = context.WithCancel(context.Background())
+
+	return sp
+}
+
+func (sp *ScrapePool) Stop() {
+	if sp.cancel != nil {
+		sp.cancel()
+	}
+}
+
+// Sync the running scrapers with the provided list of targets.
+func (sp *ScrapePool) Sync(targets []*Target) {
+	sp.mtx.Lock()
+	defer sp.mtx.Unlock()
+
+	// Shutdown all running loops after their current scrape
```

Comment:
That's exactly the point of all of this. We want to keep this simple.

What happens here is that we stop all scrape loops. They do not terminate until they have successfully finished any ongoing scrapes.
Before those terminate we already start all the new loops. With predictable slotting the first scrape of the new loops will be exactly the right interval after the last scrape of the old loops.

There is a tiny gap between *initiating* all the stops and immediately starting all the new loops. Should this really be an issue, i.e. targets frequently missing scrapes, we can place a timer and let the new loops start scrapes the fell into the gap anyway.

Simplifying the design here for a few more irrelevant allocations and cycles is well worth it.